NewHMAC panics if the HMAC doesn't support a valid digest

The `hmac.New` function falls back to Go crypto when it is called with a hash function not supported by BoringCrypto's HMAC implementation. This is is signaled by making the bindings return a nil hasher: https://github.com/golang/go/blob/fc9f02c7aec81bcfcc95434d2529e0bb0bc03d66/src/crypto/hmac/hmac.go#L135

We tried to replicate the same behavior in `openssl.NewHMAC`:

https://github.com/golang-fips/openssl/blob/f8eea431f613cac3064280dbb97d9ce1a39de973/hmac.go#L25-L30

The problem is that some OpenSSL providers return a valid hasher that can't be used as a HMAC hasher. An example is the [SymCrypt-OpenSSL](https://github.com/microsoft/SymCrypt-OpenSSL) provider, which doesn't implement HMAC-SHA224 (see test failure below).

We should instruct `openssl.NewHMAC` to return a nil HMAC object when that happens. 

```cmd
go test -run TestHMAC/sha224
Using libcrypto.so.3
OpenSSL version: OpenSSL 3.3.0 9 Apr 2024
FIPS enabled: false
[ERROR] error:41000066:SCOSSL::Algorithm not supported by SCOSSL:SCOSSL does not support hash algorithm for MAC 675 at /usr/src/azl/BUILD/SymCrypt-OpenSSL-1.5.1/ScosslCommon/src/scossl_mac.c, line 193
--- FAIL: TestHMAC (0.00s)
    --- FAIL: TestHMAC/sha224 (0.00s)
panic: EVP_MAC_init failed
openssl error(s):
error:41000066:SCOSSL::Algorithm not supported by SCOSSL
        /usr/src/azl/BUILD/SymCrypt-OpenSSL-1.5.1/ScosslCommon/src/scossl_mac.c:193
error:1C80007D:Provider routines::invalid mode
        /usr/src/azl/BUILD/SymCrypt-OpenSSL-1.5.1/SymCryptProvider/src/mac/p_scossl_hmac.c:164 [recovered]
        panic: EVP_MAC_init failed
openssl error(s):
error:41000066:SCOSSL::Algorithm not supported by SCOSSL
        /usr/src/azl/BUILD/SymCrypt-OpenSSL-1.5.1/ScosslCommon/src/scossl_mac.c:193
error:1C80007D:Provider routines::invalid mode
        /usr/src/azl/BUILD/SymCrypt-OpenSSL-1.5.1/SymCryptProvider/src/mac/p_scossl_hmac.c:164

goroutine 7 [running]:
testing.tRunner.func1.2({0x5b1560, 0xc00002e5b0})
        /usr/lib/golang/src/testing/testing.go:1631 +0x24a
testing.tRunner.func1()
        /usr/lib/golang/src/testing/testing.go:1634 +0x377
panic({0x5b1560?, 0xc00002e5b0?})
        /usr/lib/golang/src/runtime/panic.go:770 +0x132
github.com/golang-fips/openssl/v2.newHMAC3({0xc000024080, 0x40, 0x40}, {0x63a208, 0xc00010aa50}, 0x7f44a0005bc0)
        /root/openssl/hmac.go:113 +0x345
github.com/golang-fips/openssl/v2.NewHMAC(0xc000118820?, {0x0, 0x0, 0x0})
        /root/openssl/hmac.go:45 +0xac
github.com/golang-fips/openssl/v2.TestHMAC.func1(0xc000118820)
        /root/openssl/hmac_test.go:24 +0x4a
testing.tRunner(0xc000118820, 0xc000038140)
        /usr/lib/golang/src/testing/testing.go:1689 +0xfb
created by testing.(*T).Run in goroutine 6
        /usr/lib/golang/src/testing/testing.go:1742 +0x390
exit status 2
FAIL    github.com/golang-fips/openssl/v2       0.009s
```

--- a/hmac.go
+++ b/hmac.go
@@ -13,11 +13,6 @@ import (
 
 var OSSL_MAC_PARAM_DIGEST = C.CString("digest")
 
-var (
-	fetchHMACOnce sync.Once
-	evpHMAC       C.GO_EVP_MAC_PTR
-)
-
 // NewHMAC returns a new HMAC using OpenSSL.
 // The function h must return a hash implemented by
 // OpenSSL (for example, h could be openssl.NewSHA256).
@@ -38,14 +33,29 @@ func NewHMAC(h func() hash.Hash, key []byte) hash.Hash {
 		key = make([]byte, C.GO_EVP_MAX_MD_SIZE)
 	}
 
+	hmac := &opensslHMAC{
+		size:      ch.Size(),
+		blockSize: ch.BlockSize(),
+	}
+
 	switch vMajor {
 	case 1:
-		return newHMAC1(key, ch, md)
+		ctx := newHMAC1(key, md)
+		if ctx.ctx == nil {
+			return nil
+		}
+		hmac.ctx1 = ctx
 	case 3:
-		return newHMAC3(key, ch, md)
+		ctx := newHMAC3(key, md)
+		if ctx.ctx == nil {
+			return nil
+		}
+		hmac.ctx3 = ctx
 	default:
 		panic(errUnsupportedVersion())
 	}
+	runtime.SetFinalizer(hmac, (*opensslHMAC).finalize)
+	return hmac
 }
 
 // hmacCtx3 is used for OpenSSL 1.
@@ -67,7 +77,7 @@ type opensslHMAC struct {
 	sum       []byte
 }
 
-func newHMAC1(key []byte, h hash.Hash, md C.GO_EVP_MD_PTR) *opensslHMAC {
+func newHMAC1(key []byte, md C.GO_EVP_MD_PTR) hmacCtx1 {
 	ctx := hmacCtxNew()
 	if ctx == nil {
 		panic("openssl: EVP_MAC_CTX_new failed")
@@ -75,42 +85,76 @@ func newHMAC1(key []byte, h hash.Hash, md C.GO_EVP_MD_PTR) *opensslHMAC {
 	if C.go_openssl_HMAC_Init_ex(ctx, unsafe.Pointer(&key[0]), C.int(len(key)), md, nil) == 0 {
 		panic(newOpenSSLError("HMAC_Init_ex failed"))
 	}
-	hmac := &opensslHMAC{
-		size:      h.Size(),
-		blockSize: h.BlockSize(),
-		ctx1:      hmacCtx1{ctx},
-	}
-	runtime.SetFinalizer(hmac, (*opensslHMAC).finalize)
-	return hmac
+	return hmacCtx1{ctx}
 }
 
-func newHMAC3(key []byte, h hash.Hash, md C.GO_EVP_MD_PTR) *opensslHMAC {
-	fetchHMACOnce.Do(func() {
-		name := C.CString("HMAC")
-		evpHMAC = C.go_openssl_EVP_MAC_fetch(nil, name, nil)
-		C.free(unsafe.Pointer(name))
-	})
-	if evpHMAC == nil {
+var hmacDigestsSupported sync.Map
+var fetchHMAC3 = sync.OnceValue(func() C.GO_EVP_MAC_PTR {
+	name := C.CString("HMAC")
+	mac := C.go_openssl_EVP_MAC_fetch(nil, name, nil)
+	C.free(unsafe.Pointer(name))
+	if mac == nil {
 		panic("openssl: HMAC not supported")
 	}
-	ctx := C.go_openssl_EVP_MAC_CTX_new(evpHMAC)
-	if ctx == nil {
-		panic("openssl: EVP_MAC_CTX_new failed")
-	}
-	digest := C.go_openssl_EVP_MD_get0_name(md)
+	return mac
+})
+
+func buildHMAC3Params(digest *C.char) C.GO_OSSL_PARAM_PTR {
 	bld := C.go_openssl_OSSL_PARAM_BLD_new()
 	if bld == nil {
 		panic(newOpenSSLError("OSSL_PARAM_BLD_new"))
 	}
 	defer C.go_openssl_OSSL_PARAM_BLD_free(bld)
 	C.go_openssl_OSSL_PARAM_BLD_push_utf8_string(bld, OSSL_MAC_PARAM_DIGEST, digest, 0)
-	params := C.go_openssl_OSSL_PARAM_BLD_to_param(bld)
+	return C.go_openssl_OSSL_PARAM_BLD_to_param(bld)
+}
+
+func isHMAC3DigestSupported(digest string) bool {
+	if _, ok := hmacDigestsSupported.Load(digest); ok {
+		return true
+	}
+	ctx := C.go_openssl_EVP_MAC_CTX_new(fetchHMAC3())
+	if ctx == nil {
+		panic(newOpenSSLError("EVP_MAC_CTX_new"))
+	}
+	defer C.go_openssl_EVP_MAC_CTX_free(ctx)
+
+	cdigest := C.CString(digest)
+	defer C.free(unsafe.Pointer(cdigest))
+	params := buildHMAC3Params(cdigest)
 	if params == nil {
 		panic(newOpenSSLError("OSSL_PARAM_BLD_to_param"))
 	}
 	defer C.go_openssl_OSSL_PARAM_free(params)
+
+	supported := C.go_openssl_EVP_MAC_CTX_set_params(ctx, params) != 0
+	hmacDigestsSupported.Store(digest, supported)
+	return supported
+}
+
+func newHMAC3(key []byte, md C.GO_EVP_MD_PTR) hmacCtx3 {
+	digest := C.go_openssl_EVP_MD_get0_name(md)
+	if !isHMAC3DigestSupported(C.GoString(digest)) {
+		// The digest is not supported by the HMAC provider.
+		// Don't panic here so the Go standard library to
+		// fall back to the Go implementation.
+		// See https://github.com/golang-fips/openssl/issues/153.
+		return hmacCtx3{}
+	}
+	params := buildHMAC3Params(digest)
+	if params == nil {
+		panic(newOpenSSLError("OSSL_PARAM_BLD_to_param"))
+	}
+	defer C.go_openssl_OSSL_PARAM_free(params)
+
+	ctx := C.go_openssl_EVP_MAC_CTX_new(fetchHMAC3())
+	if ctx == nil {
+		panic(newOpenSSLError("EVP_MAC_CTX_new"))
+	}
+
 	if C.go_openssl_EVP_MAC_init(ctx, base(key), C.size_t(len(key)), params) == 0 {
-		panic(newOpenSSLError("EVP_MAC_init failed"))
+		C.go_openssl_EVP_MAC_CTX_free(ctx)
+		panic(newOpenSSLError("EVP_MAC_init"))
 	}
 	var hkey []byte
 	if vMinor == 0 && vPatch <= 2 {
@@ -122,13 +166,7 @@ func newHMAC3(key []byte, h hash.Hash, md C.GO_EVP_MD_PTR) *opensslHMAC {
 		hkey = make([]byte, len(key))
 		copy(hkey, key)
 	}
-	hmac := &opensslHMAC{
-		size:      h.Size(),
-		blockSize: h.BlockSize(),
-		ctx3:      hmacCtx3{ctx, hkey},
-	}
-	runtime.SetFinalizer(hmac, (*opensslHMAC).finalize)
-	return hmac
+	return hmacCtx3{ctx, hkey}
 }
 
 func (h *opensslHMAC) Reset() {

--- a/hmac.go
+++ b/hmac.go
@@ -110,8 +110,8 @@ func buildHMAC3Params(digest *C.char) C.GO_OSSL_PARAM_PTR {
 }
 
 func isHMAC3DigestSupported(digest string) bool {
-	if _, ok := hmacDigestsSupported.Load(digest); ok {
-		return true
+	if v, ok := hmacDigestsSupported.Load(digest); ok {
+		return v.(bool)
 	}
 	ctx := C.go_openssl_EVP_MAC_CTX_new(fetchHMAC3())
 	if ctx == nil {

--- a/shims.h
+++ b/shims.h
@@ -344,6 +344,7 @@ DEFINEFUNC(GO_EC_GROUP_PTR, EC_GROUP_new_by_curve_name, (int nid), (nid)) \
 DEFINEFUNC(void, EC_GROUP_free, (GO_EC_GROUP_PTR group), (group)) \
 DEFINEFUNC_3_0(GO_EVP_MAC_PTR, EVP_MAC_fetch, (GO_OSSL_LIB_CTX_PTR ctx, const char *algorithm, const char *properties), (ctx, algorithm, properties)) \
 DEFINEFUNC_3_0(GO_EVP_MAC_CTX_PTR, EVP_MAC_CTX_new, (GO_EVP_MAC_PTR arg0), (arg0)) \
+DEFINEFUNC_3_0(int, EVP_MAC_CTX_set_params, (GO_EVP_MAC_CTX_PTR ctx, const GO_OSSL_PARAM_PTR params), (ctx, params)) \
 DEFINEFUNC_3_0(void, EVP_MAC_CTX_free, (GO_EVP_MAC_CTX_PTR arg0), (arg0)) \
 DEFINEFUNC_3_0(GO_EVP_MAC_CTX_PTR, EVP_MAC_CTX_dup, (const GO_EVP_MAC_CTX_PTR arg0), (arg0)) \
 DEFINEFUNC_3_0(int, EVP_MAC_init, (GO_EVP_MAC_CTX_PTR ctx, const unsigned char *key, size_t keylen, const GO_OSSL_PARAM_PTR params), (ctx, key, keylen, params)) \

